kernel/cClosure.ml

```diff
@@ -520,22 +527,25 @@ let mk_lambda env t =
 
 let destFLambda clos_fun t =
   match t.term with
-      FLambda(_,[(na,ty)],b,e) -> (na,clos_fun e ty,clos_fun (subs_lift e) b)
+    FLambda(_,[(na,ty)],b,e) -> (na,clos_fun e ty,clos_fun (fstapp subs_lift e) b)
     | FLambda(n,(na,ty)::tys,b,e) ->
-        (na,clos_fun e ty,{norm=Cstr;term=FLambda(n-1,tys,b,subs_lift e)})
+      (na,clos_fun e ty,{norm=Cstr;term=FLambda(n-1,tys,b,fstapp subs_lift e)})
     | _ -> assert false
 	(* t must be a FLambda and binding list cannot be empty *)
 
 (* Optimization: do not enclose variables in a closure.
    Makes variable access much faster *)
-let mk_clos e t =
+let subst_if_not_empty subst u = if Univ.Instance.is_empty subst then u else Univ.subst_instance_instance subst u
```

Comment:
This should probably be included in the Univ function.

Reply:
I guess subst_instance_instance subst u != u (using physical equality), but if the code is observing that we have a problem.

Reply:
Actually, subst_instance_instance uses a smart array map, so they should even be physically equal. Except if you call subst_instance_instance with an empty substitution while the instance has some free universe variables. Then you get an array out of bounds exception. Could your code be doing that?

Reply:
Ok, I see what you're doing now. When getting a constant body from the cache you pass Instance.empty so no universe substitution happens yet. Then you do the delayed substitutions with the actual requested instance. If some case of subst_instance_instance raises then it must be because somewhere you lost a substitution. I think you should define generalized compose_subs and lift_subs combinators to avoid all those fst / snd projections, that would make the change clearer.